internal/proxy/providers/sso_test.go

```diff
@@ -271,6 +304,100 @@ func TestSSOProviderGetEmailAddress(t *testing.T) {
+		Name                    string
+		Code                    string
+		ExpectedError           string
+		RedeemResponse          *redeemResponse
+		RedeemResponseInternal  *redeemResponse
+		ProfileResponse         *profileResponse
+		ProfileResponseInternal *profileResponse
+	}{
+		{
+			Name:          "redeem fails without code",
+			ExpectedError: "missing code",
+		},
+		{
+			Name:          "redeem fails if redemption server not responding",
+			Code:          "code1234",
+			ExpectedError: "got 400",
+		},
+		{
+			Name: "redeem successful",
+			Code: "code1234",
+			RedeemResponse: &redeemResponse{
+				AccessToken:  "a1234",
+				ExpiresIn:    10,
+				RefreshToken: "refresh12345",
+				Email:        "michael.bland@gsa.gov",
+			},
+			RedeemResponseInternal: &redeemResponse{
+				AccessToken:  "i1234",
+				ExpiresIn:    10,
+				RefreshToken: "refreshinternal12345",
+				Email:        "michael.bland@gsa.gov",
+			},
+			ProfileResponse: &profileResponse{
+				Email:  "michael.bland@gsa.gov",
+				Groups: []string{"core@gsa.gov"},
+			},
+			ProfileResponseInternal: &profileResponse{
+				Email:  "michael.bland@gsa.gov",
+				Groups: []string{"internal@gsa.gov"},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			p := newSSOProvider()
+
+			var redeemServer *httptest.Server
+			var redeemServerInternal *httptest.Server
+			// set up redemption resource
+			if tc.RedeemResponseInternal != nil {
+				body, err := json.Marshal(tc.RedeemResponse)
+				bodyInternal, err := json.Marshal(tc.RedeemResponseInternal)
+				testutil.Equal(t, nil, err)
+				p.RedeemURL, redeemServer = newTestServer(http.StatusOK, body)
+				p.ProxyRedeemURL, redeemServerInternal = newTestServer(http.StatusOK, bodyInternal)
+			} else {
+				p.RedeemURL, redeemServer = newCodeTestServer(500)
+				p.ProxyRedeemURL, redeemServerInternal = newCodeTestServer(400)
+			}
+			defer redeemServer.Close()
+			defer redeemServerInternal.Close()
+
+			var profileServer *httptest.Server
+			var profileServerInternal *httptest.Server
+			if tc.ProfileResponseInternal != nil {
+				body, err := json.Marshal(tc.ProfileResponse)
+				bodyInternal, err := json.Marshal(tc.ProfileResponseInternal)
+				testutil.Equal(t, nil, err)
+				p.ProfileURL, profileServer = newTestServer(http.StatusOK, body)
+				p.ProxyRedeemURL, profileServerInternal = newTestServer(http.StatusOK, bodyInternal)
+			} else {
+				p.RedeemURL, profileServer = newCodeTestServer(500)
+				p.ProxyRedeemURL, profileServerInternal = newCodeTestServer(400)
+			}
+			defer profileServer.Close()
+			defer profileServerInternal.Close()
+
+			session, err := p.Redeem("http://redirect/", tc.Code)
+			if tc.RedeemResponseInternal != nil {
+				testutil.Equal(t, nil, err)
+				testutil.NotEqual(t, session, nil)
+				testutil.Equal(t, tc.RedeemResponseInternal.Email, session.Email)
+				testutil.Equal(t, tc.RedeemResponseInternal.AccessToken, session.AccessToken)
```

Comment:
i think i have a problem here getting the `internal` server to send back the correct body

Reply:
fixed it